### PR TITLE
KAFKA-14792: Race condition in LazyIndex.get()

### PR DESCRIPTION
**Pull request by ijuma**
`LazyIndex.get()` has a race condition that can result in a ClassCastException being thrown in some cases.

This was introduced when `LazyIndex` was rewritten from Scala to Java.

I didn't include a test since it's a bit overkill to add a concurrent test for this.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LazyIndex.java [hunk 1]
```diff
@@ -166,20 +166,25 @@ public File file() {
...
+            return ((IndexValue<T>) wrapper).index;
+        else {
             lock.lock();
             try {
-                IndexFile indexFile = (IndexFile) indexWrapper;
-                IndexValue<T> indexValue = new IndexValue<>(loadIndex(indexFile.file));
-                indexWrapper = indexValue;
-                return indexValue.index;
+                if (indexWrapper instanceof IndexValue<?>)
+                    return ((IndexValue<T>) indexWrapper).index;
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LazyIndex.java:176]: The bug is that we did not check inside the lock if `indexWrapper` was of type `IndexValue`.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LazyIndex.java [hunk 2]
```diff
@@ -166,20 +166,25 @@ public File file() {
 
     @SuppressWarnings("unchecked")
     public T get() throws IOException {
-        if (indexWrapper instanceof IndexValue<?>)
-            return ((IndexValue<T>) indexWrapper).index;
-        else if (indexWrapper instanceof IndexFile) {
+        IndexWrapper wrapper = indexWrapper;
+        if (wrapper instanceof IndexValue<?>)
+            return ((IndexValue<T>) wrapper).index;
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LazyIndex.java:171]: This change is not strictly necessary with the way the code behaves today (we never change from `IndexValue` to something else), but it's cleaner not to access the field again given that we are doing it without a lock here.

**Review by junrao — APPROVED** [review]
@ijuma : Thanks for the PR. Good catch! LGTM

**Comment by ijuma** [comment]
2 builds passed, 1 failed with 2 unrelated tests:

> Build / JDK 17 and Scala 2.13 / org.apache.kafka.clients.consumer.internals.CooperativeConsumerCoordinatorTest.testOutdatedCoordinatorAssignment()	84 ms	1
> Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.integration.ExactlyOnceSourceIntegrationTest.testConnectorReconfiguration

**Pull request merged**